### ISSUE TITLE
setting on imagenet

### ISSUE THREAD

**Issue by mountain111** [tianyic/only_train_once#18]
According to the optimizer setting for resnet18 on cifar10 in tutorials, I apply the following setting for resnet50 on imagenet
optimizer = oto.dhspg(
    variant='sgd', 
    lr=0.1, 
    target_group_sparsity=0.4,
    weight_decay=1e-4,
    start_pruning_steps=50 * len(trainloader), # start pruning after 50 epochs
    epsilon=0.95)

However, I get a lower accuracy of about 68.  What should I change for resnet50 on imagenet? e.g., epsilon, starting_pruning_steps?

**Comment by tianyic** [comment]
@mountain111  Thanks for reaching out. The acc 68 is low that seems like you started from scratch? If so, I recommend to start from the pretrained resnet50, which is a common setup for other pruning works as well. In general, OTOv2 is flexible to start from scratch or pretrained checkpoint. 

Meanwhile, `epsilon` does not matter if group sparsity reached desired level.  `starting_pruning_steps` should be set as 15 or 30 since the lr decay period for imagenet is 30. 

More detailed tutorials including this experiment will be released this summer since we are currently busy on time-sensitive projects.

**Comment by mountain111** [comment]
Thanks for your response.
pretrained model with accuracy 76.15, 
lr=0.1,  
starting_pruning_steps=30, 
8GPUs,
epochs=120.
When the above setups are applied, the accuracy of 74.17 is attined, and still lower than the reported around 75.5.

The reproduced result is confusing me. What should I do to improve it?

Thanks!

**Comment by tianyic** [comment]
@mountain111 

Glad that you have boosted the accuracy via starting from the pretrained checkpoint. Your reproduced result has about 1.1% gap between the reported, which could be recapped via more training tricks such as data augmentation and smooth labeling. The imagenet training indeed requires a bit sophiscated training pipeline to get desired result 75%+. But in my memory, the results under 60% and 70% group sparsity are not that sensitive for the tricks. You could try them to see the difference. 

It is a bit hard to precisely point out where to start without the code. As mentioned, we will release more tutorials including this experiment once we are free from time-sensitive projects this summer. At the present, you could share your training script for me tiachen@microsoft.com, I will take a look if I have bandwidth but frankly not guarantee at least by the early of June.

**Comment by mountain111** [comment]
@tianyic  Thanks for your quick reply!

The code I apply is referred to https://github.com/pytorch/examples/blob/main/imagenet/main.py, where _normalization_,
 _transforms.RandomResizedCrop(224)_, and _transforms.RandomHorizontalFlip()_ are adopted.

**Comment by mountain111** [comment]
@tianyic  Hi, Tianyi
Could you please tell me the exact accuracies for 50%/60%/70% sparsity with OTOv2?

Look forward to your reply. 

Thanks!

**Comment by tianyic** [comment]
@mountain111 

Good question. The best results were 74.5, 72.3 and 70.3, which were actually reported in a table. But it was then replaced by a figure for better visualization upon reviewer.

Note that when we release the tutorials this summer, we will update the arxiv version accordingly upon the latest OTO library for the ease of comparison for future works. Since the OTO library is actively evolving, some hyper-parameters and mechanisms have been simplified for the ease of usage, then the arxiv version may have negligible difference. 

Before the coming tutorial release, I would recommend to read the [tricks for imagenet](https://assets.amazon.science/74/02/121df5234881bd2a9bca512a8f1e/bag-of-tricks-for-image-classification.pdf) if you haven't. I have not went through https://github.com/pytorch/examples/blob/main/imagenet/main.py in depth due to my workloads, while it indeed looks relatively simple, I am not sure if you could use it to reproduce 76.1 acc on full resnet50.

**Comment by tianyic** [comment]
Closed as the tutorial has been uploaded. We will update the arxiv version with the concrete numbers for the ease of comparison in this month later.
